docs/design/NsfsVersioning.md

```diff
@@ -0,0 +1,92 @@
+# NSFS Versioning
+
+## OVERVIEW
+NooBaa NSFS versioning feature used for keeping multiple variants of an object in the same NSFS bucket. A user can use the NSFS Versioning feature to preserve, retrieve, and restore every version of every object stored in a versioning enabled bucket. S3 WORM locking feature is based on versioned enabled buckets.
+
+
+## GOALS
+* Maintain a directory tree of the NSFS bucket 
+* Support S3 versioning API for NSFS buckets -
+    * Bucket API - 
+        * Get bucket versioning mode 
+        * Set bucket versioning mode
+    * Object API - 
+        * List objects
+        * List object versions
+        * Put object/Complete multipart upload
+        * Delete object/Delete multiple object of latest version/a given version
+        * Get object/Head object of latest version/a given version
+    
+
+## SOLUTION
+The Versioning implementation consists of lazy creation of a hidden sub directory that will contain all versions of all objects under the parent directory except of the latest version. 
+The latest version will be located in the parent directory itself.
+
+
+### PROPOSED SOLUTION DIAGRAM
+The following diagram illustrates a directory tree of a versioning enabled bucket
+
+<div id="top" />
+<img src="/docs/design/images/nsfs_versioning_dir_tree.png" />
+In the figure, each "regular" directory contains a hidden .versions/ sub directory that stores past versions of objects located under the parent directory, while the latest version of each object can be found under the parent directory itself. For instance, the latest version of Obj1 can be found under the root directory (/), while the old versions of Obj1 reside under /.versions/.
+
+
+## CONCEPTS 
+
+* Objects that are stored in a bucket before you set the versioning state have a version ID of null. 
+* Objects that are stored in a bucket After you set the versioning state have a unique version ID. 
+* PUT, POST and COPY operations will create a new version of the object identified by a unique version id.
+* DELETE latest version will create a delete marker which is a dummy version identified by a  unique version id as well.
+* DELETE version id will delete the version completely.
+
+
+### Version ID 
+* Will be allocated to a version on PUT/POST request (or DELETE version id). 
+* Consisted of the version mtimeMS + inode number.
+* The version ID will be attached as an xattr of the file.
+* When latest versions is moved to past versions directory, the file name will be changed from key to key_{version_id}.
```

Comment:
```suggestion
* The version ID will be attached as a xattr of the file.
* When the latest version is moved to the past versions directory, the file name will be changed from key to key_{version_id}.
```

Reply:
https://www.pristineword.com/grammar-an-x/#:~:text=It's%20an%20X.,the%20indefinite%20article%20%22an%22.